<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="manta">

    <xacro:property name="stl_scale" value="0.01" />

    <!-- This will include all the materials (colors) used to describe the manta -->
    <xacro:include filename="materials.xacro" />


    <link name="base_link">
    </link>

    
    <joint name="imu_joint" type="fixed">
        <origin xyz="0 0.5 0.3" rpy="${pi} 0 0"/>
        <parent link="base_link"/>
        <child link="imu_link"/>        
    </joint>

    <link name="imu_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.2 0.3 0.01"/>
            </geometry>
            <material name="red" />
        </visual>
    </link>

    <joint name="imu_joint_enu" type="fixed">
        <origin xyz="0 0 0" rpy="${pi} 0 ${pi/2}"/>
        <parent link="imu_link"/>
        <child link="imu_link_enu"/>        
    </joint>

    <link name="imu_link_enu">
    </link>

    <joint name="bar_joint" type="fixed">
        <origin xyz="0.2 0 0.3" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="bar_link"/>        
    </joint>

    <link name="bar_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.2" radius="0.1"/>
            </geometry>
            <material name="black" />
        </visual>
    </link>

    <joint name="gps_joint" type="fixed">
        <origin xyz="0 -0.5 0.3" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="gps_link"/>        
    </joint>

    <link name="gps_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.2 0.3 0.01"/>
            </geometry>
            <material name="green" />
        </visual>
    </link>

    <link name="base_body">
        <visual>
            <origin xyz="0 0 -0.25" rpy="0 0 0"/>
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/Fixed_body.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="yellow" />
        </visual>
    </link>

    <joint name="base_joint" type="fixed">
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="base_body"/>        
    </joint>

    <joint name="finl_joint" type="revolute">
        <origin xyz="-0.9 0.8 0.05" rpy="${pi/2} 0 ${pi}"/>
        <parent link="base_link"/>
        <child link="finl_link"/>
        <axis xyz="0 0 1"/>
        <limit lower="${-pi/2}" upper="${pi/2}" velocity="100" effort="100"/> 
        <dynamics damping="0.0" friction="0.0"/>
    </joint>

    <link name="finl_link">
        <visual>
            <origin xyz="-0.92 0.3 -0.8" rpy="${pi/2} ${pi} 0"/>   
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/finL.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="yellow" />
        </visual>
    </link>

    <joint name="finr_joint" type="revolute">
        <origin xyz="-0.9 -0.8 0.05" rpy="${-pi/2} 0 ${pi}"/>
        <parent link="base_link"/>
        <child link="finr_link"/>
        <axis xyz="0 0 1"/>
        <limit lower="${-pi/2}" upper="${pi/2}" velocity="100" effort="100"/> 
        <dynamics damping="0.0" friction="0.0"/>
    </joint>

    <link name="finr_link">
        <visual>
            <origin xyz="-0.92 -0.3 -0.8" rpy="-${pi/2} ${pi} 0"/>    
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/finR.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="yellow" />
        </visual>
    </link>

    <joint name="wingl_joint" type="revolute">
        <origin xyz="0.73 0.65 0" rpy="${pi/2} 0 ${pi/2}"/>
        <parent link="base_link"/>
        <child link="wingl_link"/>
        <axis xyz="0 0 1"/>
        <limit lower="${-pi/2}" upper="${pi/2}" velocity="100" effort="100"/> 
        <dynamics damping="0.0" friction="0.0"/>
    </joint>

    <link name="wingl_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>   
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/wing_left.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="blue" />
        </visual>
    </link>

    <joint name="wingr_joint" type="revolute">
        <origin xyz="0.73 -0.65 0" rpy=" ${-pi/2} 0 ${-pi/2}"/>
        <parent link="base_link"/>
        <child link="wingr_link"/>
        <axis xyz="0 0 1"/>
        <limit lower="${-pi/2}" upper="${pi/2}" velocity="100" effort="100"/> 
        <dynamics damping="0.0" friction="0.0"/>
    </joint>

    <link name="wingr_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>   
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/wing_right.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="blue" />
        </visual>
    </link>

    <joint name="cameraL_joint" type="fixed">
        <parent link="base_link"/>
        <child link="cameraL_link"/>
        <origin xyz="1.1 0.6 -0.05" rpy="0 ${3*pi/4} 0"/>
    </joint>

    <link name="cameraL_link">
        <visual>
            <origin xyz="-0.1 0.05 0.1" rpy="0 ${pi} 0"/>   
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/RaspiCamV3.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="black" />
        </visual>
    </link>

    <joint name="cameraR_joint" type="fixed">
        <parent link="base_link"/>
        <child link="cameraR_link"/>
        <origin xyz="1.1 -0.6 -0.05" rpy="0 ${3*pi/4} 0"/>
    </joint>

    <link name="cameraR_link">
        <visual>
            <origin xyz="0.1 -0.05 0.1" rpy="0 ${pi} ${pi}"/>   
            <geometry>
                <mesh filename="https://raw.githubusercontent.com/ScudeT/public_files/manta_files/RaspiCamV3.stl" scale="${stl_scale} ${stl_scale} ${stl_scale}"/>
            </geometry>
            <material name="black" />
        </visual>
    </link>

</robot>